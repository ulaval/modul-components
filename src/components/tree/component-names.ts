export const TREE_NODE_NAME: string = 'm-tree-node';
export const TREE_ICON_NAME: string = 'm-tree-icon';
