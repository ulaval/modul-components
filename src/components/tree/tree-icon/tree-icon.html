<m-icon class="m-tree-icon"
        v-if="showFolderIcon"
        size="24px"
        :showNameAsClass="true"
        :name="folderIcon"></m-icon>
<m-icon-file class="m-tree-icon"
             v-else-if="useFilesIcons"
             :name="extensionFile"
             :extension="extensionFile"></m-icon-file>
<m-plus class="m-tree-icon"
        v-else-if="isFolder"
        ref="plus-icon"
        :class="{ 'm--is-left': 'true' }"
        :open="isFolderOpen"
        :large="false"
        skin="light"
        aria-hidden="true"></m-plus>
