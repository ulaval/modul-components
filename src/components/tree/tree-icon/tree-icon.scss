@import 'commons';

.m-tree-icon {
    margin-right: $m-spacing--s;
    min-width: 32px;

    /deep/ .m-svg__folder-open {
        transform: scale(1.2) translateY(-2px);
    }
}
