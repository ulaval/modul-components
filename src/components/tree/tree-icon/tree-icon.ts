import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import { extractExtension } from '../../../utils/file/file';
import { ModulVue } from '../../../utils/vue/vue';
import WithRender from './tree-icon.html?style=./tree-icon.scss';

const FOLDER_OPEN: string = 'm-svg__folder-open';
const FOLDER_CLOSED: string = 'm-svg__folder';

@WithRender
@Component
export class MTreeIcon extends ModulVue {
    @Prop()
    public filename: string;

    @Prop()
    public isFolderOpen: boolean;

    @Prop({ default: false })
    public isFolder: boolean;

    @Prop({ default: false })
    public useFilesIcons: boolean;

    public get folderIcon(): string {
        return this.isFolderOpen ? FOLDER_OPEN : FOLDER_CLOSED;
    }

    public get extensionFile(): string {
        return '.' + extractExtension(this.filename);
    }

    public get showFolderIcon(): boolean {
        return this.isFolder && this.useFilesIcons;
    }
}

