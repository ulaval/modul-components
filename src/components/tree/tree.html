<div class="m-tree">
    <m-message v-if="errorTree" class="m-tree__error" state="warning" skin="light" ref="error-tree-txt">
        {{'m-tree:error' | f-m-i18n}}
    </m-message>
    <p v-if="propTreeEmpty" class="m-tree__empty" ref="empty-tree-txt">
        {{'m-tree:empty' | f-m-i18n}}
    </p>
    <ul v-else class="m-tree__content">
        <m-tree-node ref="tree-node" v-for="node in tree" :key="node.id" :selectable="selectable" :open.sync="node.open" :node="node" :selected-nodes="propSelectedNodes"
            :icons="icons" @click="onClick"></m-tree-node>
    </ul>
</div>
