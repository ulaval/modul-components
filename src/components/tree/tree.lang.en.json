{
    "m-tree": {
        "empty": "No content",
        "error": "Some items could not be displayed because of a technical problem."
    }
}
