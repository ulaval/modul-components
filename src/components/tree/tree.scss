@import '../../styles/abstracts/variables';

.m-tree {
    overflow: auto;

    &__error {
        .m-icon {
            color: $m-color--error-dark;
        }

        + .m-tree__content {
            margin-top: $m-margin;
        }
    }

    &__content {
        margin-top: 0;
        margin-bottom: 0;
    }
}
