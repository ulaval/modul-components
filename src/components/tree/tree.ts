import { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Prop, Watch } from 'vue-property-decorator';

import { ModulVue } from '../../utils/vue/vue';
import { TREE_NAME } from '../component-names';
import I18nPlugin from '../i18n/i18n';
import TreeNodePlugin from '../tree-node/tree-node';
import WithRender from './tree.html?style=./tree.scss';

export interface TreeNode {
    id: string;
    label?: string;
    open?: boolean;
    children?: TreeNode[];
    hasChildren?: boolean;
    data?: any;
}

export enum MSelectionMode {
    None = '0',
    Single = '1',
    Multiple = '2'
}

@WithRender
@Component
export class MTree extends ModulVue {
    @Prop()
    public tree: TreeNode[];

    @Prop({
        default: MSelectionMode.Single,
        validator: value =>
            value === MSelectionMode.None ||
            value === MSelectionMode.Single ||
            value === MSelectionMode.Multiple
    })
    public selectionMode: MSelectionMode;

    @Prop()
    public selectedNodes: string[];

    @Prop()
    public icons: boolean;

    public propSelectedNodes: string[] = this.selectedNodes || [];
    public errorTree: boolean = false;

    private selectedNodesFound: string[] = [];

    public onClick(path: string): void {
        if (this.propSelectedNodes.indexOf(path) === -1) {
            if (this.selectionMode === MSelectionMode.Multiple) {
                this.propSelectedNodes.push(path);
            } else {
                this.propSelectedNodes = [path];
            }
        } else if (this.selectionMode === MSelectionMode.Multiple) {
            this.propSelectedNodes.splice(this.propSelectedNodes.indexOf(path), 1);
        }
        this.$emit('select', path);
    }

    public get propTreeEmpty(): boolean {
        return !this.tree.length;
    }

    public get selectable(): boolean {
        return this.selectionMode !== MSelectionMode.None;
    }

    protected created(): void {
        this.browseTree();
    }

    @Watch('tree')
    private browseTree(): void {
        this.errorTree = false;
        this.tree.forEach(node => {
            this.browseNode(node);
        });
        this.propSelectedNodes.forEach(selectedNode => {
            if (this.selectedNodesFound.indexOf(selectedNode) === -1) {
                console.error(`modUL - The selected node was not found: "${selectedNode}"`);
            }
        });
    }

    private browseNode(node: TreeNode, path: string = ''): void {
        if (node.id.trim() === '') {
            this.errorTree = true;
        }
        let currentPath: string = path + '/' + node.id;
        if (this.propSelectedNodes.indexOf(currentPath) !== -1) {
            this.selectedNodesFound.push(currentPath);
        }
        if (node.children) {
            node.children.forEach(childNode => {
                this.browseNode(childNode, currentPath);
            });
        }
    }
}

const TreePlugin: PluginObject<any> = {
    install(v, options): void {
        v.prototype.$log.debug(TREE_NAME, 'plugin.install');
        v.use(TreeNodePlugin);
        v.use(I18nPlugin);
        v.component(TREE_NAME, MTree);
    }
};

export default TreePlugin;
