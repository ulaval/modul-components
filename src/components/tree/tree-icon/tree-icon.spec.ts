import { RefSelector, shallowMount, Wrapper } from '@vue/test-utils';
import { TreeNode } from '../tree';
import { MTreeIcon } from './tree-icon';

const TREE_NODE_FOLDER: TreeNode = {
    label: 'Node 2',
    id: '/Node 2'
};

const FOLDER_OPEN: string = 'm-svg__folder-open';
const FOLDER_CLOSED: string = 'm-svg__folder';
const PLUS: RefSelector = { ref: 'plus-icon' };

let file: TreeNode;
let isFolderOpen: boolean = false;
let isFolder: boolean = false;
let useFilesIcons: boolean = false;
let wrapper: Wrapper<MTreeIcon>;

const initializeShallowWrapper: any = () => {
    wrapper = shallowMount(MTreeIcon, {
        stubs: getStubs(),
        propsData: {
            file,
            isFolderOpen,
            isFolder,
            useFilesIcons
        }
    });
};

const getStubs: any = () => {
    return {
        ['m-icon']: '<span>m-icon</span>',
        ['m-icon-file']: '<span>m-icon-file</span>'
    };
};

describe(`MTreeIcon`, () => {

    describe(`When the node is a folder`, () => {

        beforeEach(() => {
            file = TREE_NODE_FOLDER;
            useFilesIcons = false;
            isFolderOpen = false;
            isFolder = true;
        });

        describe(`When the node uses the plus icon (not the folder icon)`, () => {
            beforeEach(() => {
                useFilesIcons = false;
                initializeShallowWrapper();
            });

            it('Should use the m-plus component', () => {
                expect(wrapper.find(PLUS).exists()).toBeTruthy();
            });

        });

        describe(`When the node uses the folder icon`, () => {

            it('Should not include the m-plus component', () => {
                useFilesIcons = true;
                initializeShallowWrapper();
                expect(wrapper.find(PLUS).exists()).toBeFalsy();

            });

            describe(`When the folder is opened`, () => {

                it(`Should be the right icon`, () => {
                    isFolderOpen = true;
                    initializeShallowWrapper();

                    expect(wrapper.vm.folderIcon).toEqual(FOLDER_OPEN);
                });

            });

            describe(`When the folder is closed`, () => {

                it(`Should be the right icon`, () => {
                    isFolderOpen = false;
                    initializeShallowWrapper();

                    expect(wrapper.vm.folderIcon).toEqual(FOLDER_CLOSED);
                });

            });

        });

    });
});
