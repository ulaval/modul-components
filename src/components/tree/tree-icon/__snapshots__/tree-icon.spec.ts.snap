// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MTreeIcon When the node is a folder When the node uses plus icons Should render correctly 1`] = `<m-plus large="true" aria-hidden="true" class="m-accordion__header-icon m--is-left"></m-plus>`;
